llvm/lib/Analysis/TypeBasedAliasAnalysis.cpp

```diff
@@ -613,12 +613,13 @@ static bool mayBeAccessToSubobjectOf(TBAAStructTagNode BaseTag,
     }
 
     if (BaseType.getNode() == SubobjectTag.getBaseType()) {
```

Comment:
I'm sorry, I can't follow the logic here at all.  What is the purpose of the code inside this if statement?

Reply:
I think I see... the old code detects whether both accesses refer to the same object, but we also care about overlapping objects.  According to the TBAA rules, objects have to be properly nested.  So the only way to get overlap is if one of the accesses is a "whole-object" access.  Therefore, this check.

LangRef says "the access type is a scalar type descriptor", which I guess isn't accurate?  Please propose a patch to fix the language there.

Reply:
LangRef doesn't cover the new aggregate-enabled TBAA format, which this patch is about. The latest plan documentation-wise I'm aware of is to wait until the new format is mature enough to be enabled by default, and then update the documentation, see <https://reviews.llvm.org/D40975#955973>. Until then it's all work-in-progress and not something supposed to be exposed to the user.